### PR TITLE
Enhance the healthcheck to provide healthz for api and check for pod status to be running. 

### PR DESCRIPTION
**Pull request by kwoodson**
Enhance the healthcheck to ensure the following:
1. The healthz check for the API returns a 200
2. The pods are in a Running state without restarts
3. The web console is available

@jim-minter, Please review

### CODE REVIEW

File: pkg/util/util.go [hunk 1]
```diff
@@ -43,3 +49,16 @@ func Template(tmpl string, f template.FuncMap, cs *acsapi.ContainerService, c *c
 
 	return b.Bytes(), nil
 }
+
+// GetKubeconfigFromV1Config takes a v1 config and returns a kubeconfig
+func GetKubeconfigFromV1Config(kc *v1.Config) (clientcmd.ClientConfig, error) {
```
Comment (kwoodson) [inline pkg/util/util.go]: As per the comment above regarding the anti-pattern, should this function move into the `healthcheck.go`?  Seems like a useful function when you don't have a kubeconfig available.

File: pkg/checks/checks.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+		if resp.StatusCode == http.StatusOK {
+			return nil
+		}
+
+		time.Sleep(time.Second)
+	}
+}
+
+// CheckStatusOfPods verifies all pods are in running state
+func CheckStatusOfPods(coreclient *coreclient.CoreV1Client, namespace string) error {
```
Comment (jim-minter) [inline pkg/checks/checks.go]: use a context here and in WaitForHTTPStatusOk and return if the context has expired.

Caller:
```
ctx, cancel := context.WithTimeout(ctx.Background(), 5 * time.Minute)
defer cancel()
f(ctx, ...)
```

Callee:
```
func f(ctx context.Context, ...) {
	...
	for {
		...
		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
		}
	}
}
```

Comment (jim-minter) [reply pkg/checks/checks.go]: (then no need for startTime, hard-coded timeout, etc)

Comment (jim-minter) [reply pkg/checks/checks.go]: How about `func WaitForPodRunning(coreclient *coreclient.CoreV1Client, namespace, name string)` and have a list of pods you want to check for now, a bit like in https://github.com/Azure/acs-engine/blob/master/test/e2e/openshift/openshift_test.go#L145-L173

File: pkg/healthcheck/healthcheck.go [hunk 3]
```diff
@@ -10,11 +10,54 @@ import (
...
+		return err
+	}
+
+	// Ensure that the pods in default are healthy
+	namespace, _, err := kubeconfig.Namespace()
+	if err != nil {
+		return err
+	}
+
+	err = checks.CheckStatusOfPods(coreclient, namespace)
```
Comment (jim-minter) [inline pkg/healthcheck/healthcheck.go]: this is where you need to iterate a list of pods to check

File: pkg/addons/addons.go [hunk 4]
```diff
@@ -287,47 +285,17 @@ func getClients() (err error) {
...
 		if err != nil {
 			return err
 		}
-
-		err = waitForHealthz()
+		transport, err := rest.TransportFor(restconfig)
+		if err != nil {
+			return err
+		}
+		err = checks.WaitForHTTPStatusOk(transport, restconfig.Host+"/heathz")
```
[the remark this answers is not in the document]
Comment (kwoodson) [reply pkg/addons/addons.go]: fixed.

**Comment by kwoodson** [comment]
The health check portion is working.  Seeing a panic when attempting to deploy inside of the sync.go:
```
+ go run cmd/sync/sync.go
panic: Get https://openshift.kwoodsonhcp1.osadev.cloud/api: EOF

goroutine 1 [running]:
main.main()
	/home/kwoodson/go/src/github.com/jim-minter/azure-helm/cmd/sync/sync.go:55 +0x49
exit status 2
```
@jim-minter, any ideas?

File: pkg/addons/addons.go [hunk 5]
```diff
@@ -220,12 +223,20 @@ func writeDB(client *client, db map[string]unstructured.Unstructured) error {
...
 		return err
 	}
 
-	if err := client.waitForHealthz(); err != nil {
+	transport, err := rest.TransportFor(restconfig)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
```
Comment (jim-minter) [inline pkg/addons/addons.go]: just use context.Background() here for now, no timeout

File: pkg/addons/client.go [hunk 6]
```diff
@@ -36,27 +32,27 @@ type client struct {
 	dryRun bool
 }
 
-func newClient(dryRun bool) (*client, error) {
+func newClient(dryRun bool) (*client, *rest.Config, error) {
```
Comment (jim-minter) [inline pkg/addons/client.go]: restconfig is already a field in the returned client object, are these changes necessary?

File: pkg/checks/checks.go [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(time.Second):
+		}
+	}
+}
+
+// WaitForPodsRunning verifies all pods are in running state
+func WaitForPodsRunning(ctx context.Context, coreclient *coreclient.CoreV1Client, namespace string) error {
```
Comment (jim-minter) [inline pkg/checks/checks.go]: namespace argument unnecessary

File: pkg/healthcheck/healthcheck.go [hunk 8]
```diff
@@ -10,11 +10,70 @@ import (
...
+	if err != nil {
+		return err
+	}
+
+	t, err := rest.TransportFor(restconfig)
+	if err != nil {
+		return err
+	}
+
+	ctxHealthz, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
```
Comment (jim-minter) [inline pkg/healthcheck/healthcheck.go]: use the passed in context.  The point of context is the caller gets to decide how long the total timeout is.

File: pkg/checks/checks.go [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+		return err
+	}
+
+	for {
+		resp, err := cli.Do(req)
+		if err, ok := err.(*url.Error); ok && (err.Timeout() || err.Err == io.EOF || err.Err == io.ErrUnexpectedEOF) {
+			time.Sleep(time.Second)
+			continue
+		}
+		if err != nil {
```
Comment (jim-minter) [inline pkg/checks/checks.go]: w.r.t. your error, try adding here:
```
if err == io.EOF {
	time.Sleep(time.Second)
	continue
}
```

Comment (jim-minter) [reply pkg/checks/checks.go]: if that doesn't help add `log.Printf("the error was a %#v\n", err)` to work out what the error actually was.

File: pkg/checks/checks.go [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+// WaitForPodsRunning verifies all pods are in running state
+func WaitForPodsRunning(ctx context.Context, coreclient *coreclient.CoreV1Client, namespace string) error {
+
+	whitelistedNamespaces := []string{
+		"openshift-",
+		"kube-",
+		"default",
+	}
+
+	pods, err := coreclient.Pods(namespace).List(metav1.ListOptions{})
```
Comment (jim-minter) [inline pkg/checks/checks.go]: we didn't get a chance to catch up today, sorry. I think you had some questions about this.  I'm expecting to see you looping through statefulsets, daemonsets, etc. using the apps client and not the core client.
you could use the clientset at https://github.com/jim-minter/azure-helm/blob/master/vendor/k8s.io/client-go/kubernetes/clientset.go#L399, that gives you nearly every k8s client there is.  Good to help developing.
For the final thing you're better off using more scoped clients (easier for testing).  In this case, https://github.com/jim-minter/azure-helm/blob/master/vendor/k8s.io/client-go/kubernetes/typed/apps/v1/apps_client.go#L63

File: pkg/addons/addons.go (hunk 5 above)
Comment (kwoodson) [reply pkg/addons/addons.go]: done.

File: pkg/addons/client.go (hunk 6 above)
Comment (kwoodson) [reply pkg/addons/client.go]: I didn't realize that.  removing changes.

File: pkg/checks/checks.go (hunk 7 above)
Comment (kwoodson) [reply pkg/checks/checks.go]: ok.  I will remove.

File: pkg/healthcheck/healthcheck.go (hunk 8 above)
Comment (kwoodson) [reply pkg/healthcheck/healthcheck.go]: Ok.

File: pkg/checks/checks.go (hunk 10 above)
Comment (kwoodson) [reply pkg/checks/checks.go]: I was able to find the statefulsets and daemonsets by using the apps_client.go you mentioned.  

My question would be, what is the advantage of getting those objects from the apps_client when the coreclient fetches the pods which are from those objects?  The owner reference on the pod has a `kind` field which tells you about the parent object.  

Is it specifically that daemonsets and statefulsets don't deploy when there are no matching selectors?  If this is the case, I'd lean to verifying there are nodes labeled appropriately so that the selectors are correct.

I am fine proceeding but would like your insight as to which is the preferred so that I understand.

Thanks @jim-minter

Comment (jim-minter) [reply pkg/checks/checks.go]: It's a mess and the answer is almost certainly that both clients are needed. I think it's probably best to iterate the daemonsets/deployments/etc and then resort to checking the pods only if necessary because the info isn't in the parent's status, but there are probably multiple ways to skin this particular cat.

File: pkg/checks/checks.go (hunk 9 above)
Comment (kwoodson) [reply pkg/checks/checks.go]: The `io.EOF` did the trick.

File: pkg/checks/checks.go [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+		case <-ctx.Done():
+			log.Printf("ctx.Done called.  Returning err")
+			return ctx.Err()
+		case <-time.After(time.Second):
+		}
+	}
+}
+
+// WaitForInfaServices verify daemonsets, statefulsets
+func WaitForInfaServices(ctx context.Context, appclient *appclient.AppsV1Client) error {
```
Comment (kwoodson) [inline pkg/checks/checks.go]: @jim-minter was this more in-line with what you had in mind?

Should I add the deeper dive (fetch pods and check status) when I find `ReadyReplicas != Replicas` with each type?

Comment (jim-minter) [reply pkg/checks/checks.go]: yes :)

Comment (jim-minter) [inline pkg/checks/checks.go]: Infra

File: pkg/checks/checks.go [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+
+	// Daemonsets
+	dsList, err := appclient.DaemonSets("").List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, ds := range dsList.Items {
```
Comment (jim-minter) [inline pkg/checks/checks.go]: for each kind, discard it if it isn't in a namespace we care about

File: pkg/checks/checks.go [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+				return err
+			}
+			if ds.Status.NumberMisscheduled > 0 {
+				return fmt.Errorf("Daemonset[%v] in Namespace[%v] has missscheduled[%v]", ds.Name, ds.Namespace, ds.Status.NumberMisscheduled)
+			}
+
+			if ds.Status.DesiredNumberScheduled == ds.Status.CurrentNumberScheduled && ds.Status.DesiredNumberScheduled == ds.Status.NumberReady {
+				break
+			}
+			time.Sleep(2 * time.Second)
```
Comment (jim-minter) [inline pkg/checks/checks.go]:
```
select {
	case <-time.After(2 * time.Second):
	case <-ctx.Done():
		return ctx.Err()
}
```

File: pkg/checks/checks.go (hunk 11 above)
Comment (kwoodson) [reply pkg/checks/checks.go]: Fixed.

**Comment by kwoodson** [comment]
@jim-minter, If you wouldn't mind taking a look.  If this level of detection is sufficient then I'll clean up and we can merge.  If you'd like a deeper dive into pods I can also call the pod function.

File: pkg/checks/checks.go [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+	}
+	for _, ns := range whitelistedNamespaces {
+		if strings.HasPrefix(namespace, ns) {
+			return true
+		}
+	}
+	return false
+}
+
+// WaitForInfraServices verify daemonsets, statefulsets
```
Comment (kargakis) [inline pkg/checks/checks.go]: We probably need a hardcoded list of the expected workloads and cross-checking readiness vs mere listing. I am fine with a follow-up issue.

File: pkg/checks/checks.go [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+			select {
+			case <-time.After(2 * time.Second):
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+		}
+	}
+
+	// ReplicaSets
+	rsList, err := appclient.ReplicaSets("").List(metav1.ListOptions{})
```
Comment (kargakis) [inline pkg/checks/checks.go]: You shouldn't need to check for replicasets since you check for deployments, unless some replicasets run standalone.

File: pkg/healthcheck/healthcheck.go [hunk 16]
```diff
@@ -10,11 +10,75 @@ import (
...
+	t, err := rest.TransportFor(restconfig)
+	if err != nil {
+		return err
+	}
+
+	ctxHealthz, cancel := context.WithTimeout(ctx, 5*time.Minute)
+	defer cancel()
+
+	// Wait for the healthz to be 200 status
+	err = checks.WaitForHTTPStatusOk(ctxHealthz, t, restconfig.Host+"/healthz")
```
Comment (kargakis) [inline pkg/healthcheck/healthcheck.go]: How about checking for `/healthz/ready` vs `/healthz`?

File: pkg/checks/checks.go (hunk 15 above)
Comment (kwoodson) [reply pkg/checks/checks.go]: Currently they do not run standalone.  Is there any harm in doing so?

File: pkg/healthcheck/healthcheck.go (hunk 16 above)
Comment (kwoodson) [reply pkg/healthcheck/healthcheck.go]: I'm not familiar with the difference.  I'll try it and see.

File: pkg/checks/checks.go [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+			log.Printf("the err was %#v\n", err)
+			return err
+		}
+		if resp.StatusCode == http.StatusOK {
+			return nil
+		}
+
+		select {
+		case <-ctx.Done():
+			log.Printf("ctx.Done called.  Returning err")
```
Comment (jim-minter) [inline pkg/checks/checks.go]: remove log.Printf?

File: pkg/checks/checks.go [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+		case <-time.After(time.Second):
+		}
+	}
+}
+
+func checkNamespace(namespace string) bool {
+	whitelistedNamespaces := []string{
+		"openshift-",
+		"kube-",
+		"default",
```
Comment (jim-minter) [inline pkg/checks/checks.go]: edge case: `default` should be tested with equality, not hasprefix

File: pkg/checks/checks.go [hunk 19]
```diff
@@ -0,0 +1,249 @@
...
+				return ctx.Err()
+			}
+		}
+	}
+
+	return nil
+}
+
+// WaitForPodsRunning verifies all pods are in running state
+func WaitForPodsRunning(ctx context.Context, coreclient *coreclient.CoreV1Client, appclient *appclient.AppsV1Client) error {
```
Comment (jim-minter) [inline pkg/checks/checks.go]: not used - remove (save in a scratch pad in case you need it one day ;)

File: pkg/healthcheck/healthcheck.go [hunk 20]
```diff
@@ -10,11 +10,75 @@ import (
...
+	if err != nil {
+		return err
+	}
+
+	// coreclient, err := coreclient.NewForConfig(restconfig)
```
Comment (jim-minter) [inline pkg/healthcheck/healthcheck.go]: remove

File: pkg/healthcheck/healthcheck.go [hunk 21]
```diff
@@ -10,11 +10,75 @@ import (
...
+
+	appsclient, err := appsclient.NewForConfig(restconfig)
+	if err != nil {
+		return err
+	}
+	// Ensure that the pods in default are healthy
+	ctxPods, cancel := context.WithTimeout(ctx, 5*time.Minute)
+	defer cancel()
+	err = checks.WaitForInfraServices(ctxPods, appsclient)
+	// err = checks.WaitForPodsRunning(ctxPods, coreclient, appsclient)
```
Comment (jim-minter) [inline pkg/healthcheck/healthcheck.go]: remove

File: pkg/checks/checks.go (hunk 17 above)
Comment (kwoodson) [reply pkg/checks/checks.go]: done.

File: pkg/checks/checks.go (hunk 19 above)
Comment (kwoodson) [reply pkg/checks/checks.go]: done

File: pkg/healthcheck/healthcheck.go (hunk 20 above)
Comment (kwoodson) [reply pkg/healthcheck/healthcheck.go]: done

File: pkg/healthcheck/healthcheck.go (hunk 21 above)
Comment (kwoodson) [reply pkg/healthcheck/healthcheck.go]: done

File: pkg/checks/checks.go (hunk 15 above)
Comment (kargakis) [reply pkg/checks/checks.go]: My initial thought was that it's redundant since the deployment status api is an aggregation of all the replicasets under its control (I wrote that part :)). Now, I think it's also wrong because you may list old replicasets that should not have available replicas running.

File: pkg/checks/checks.go [hunk 22]
```diff
@@ -0,0 +1,186 @@
...
+		if !checkNamespace(ds.Namespace) {
+			continue
+		}
+		for {
+			ds, err := appclient.DaemonSets(ds.Namespace).Get(ds.Name, metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+			if ds.Status.NumberMisscheduled > 0 {
+				return fmt.Errorf("Daemonset[%v] in Namespace[%v] has missscheduled[%v]", ds.Name, ds.Namespace, ds.Status.NumberMisscheduled)
```
Comment (kargakis) [inline pkg/checks/checks.go]: I wouldn't error out here at once but retry for some time until it converges to zero.

File: pkg/checks/checks.go (hunk 14 above)
Comment (kargakis) [reply pkg/checks/checks.go]: Yeah, maybe that's enough

Comment (kargakis) [reply pkg/checks/checks.go]: Although I am worried about a missing workload for whatever reason, that's why I think a whitelist of the workloads is eventually required.

File: pkg/checks/checks.go (hunk 22 above)
Comment (kargakis) [reply pkg/checks/checks.go]: We can reuse the existing context timeout as a retry mechanism so you can just log here.

File: pkg/checks/checks.go [hunk 23]
```diff
@@ -0,0 +1,186 @@
...
+			}
+
+			if ds.Status.DesiredNumberScheduled == ds.Status.CurrentNumberScheduled && ds.Status.DesiredNumberScheduled == ds.Status.NumberReady {
```
Comment (kargakis) [inline pkg/checks/checks.go]:
```go
if ds.Status.DesiredNumberScheduled == ds.Status.CurrentNumberScheduled && 
  ds.Status.DesiredNumberScheduled == ds.Status.NumberAvailable &&
  ds.Status.DesiredNumberScheduled == ds.Status.UpdatedNumberScheduled &&
  ds.Metadata.Generation == ds.Status.ObservedGeneration {
```

File: pkg/checks/checks.go (hunk 22 above)
Comment (kwoodson) [reply pkg/checks/checks.go]: Should I log or just remove this statement entirely and let the loop occur until we are satisfied with the valid logic?

Comment (kargakis) [reply pkg/checks/checks.go]: Removing is also probably fine. I would add a log that prints all status fields every time we retry but can be done in a follow-up across all loops if we want it.

File: pkg/checks/checks.go [hunk 24]
```diff
@@ -0,0 +1,186 @@
...
+	for _, ss := range ssList.Items {
+		if !checkNamespace(ss.Namespace) {
+			continue
+		}
+		for {
+			ss, err := appclient.StatefulSets(ss.Namespace).Get(ss.Name, metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+			if ss.Status.Replicas == ss.Status.ReadyReplicas && ss.Status.ReadyReplicas == ss.Status.CurrentReplicas {
```
[the remark this answers is not in the document]
Comment (kwoodson) [reply pkg/checks/checks.go]: Added.  Will test.

File: pkg/checks/checks.go [hunk 25]
```diff
@@ -0,0 +1,186 @@
...
+	for _, d := range dList.Items {
+		if !checkNamespace(d.Namespace) {
+			continue
+		}
+		for {
+			d, err := appclient.Deployments(d.Namespace).Get(d.Name, metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+			if d.Status.Replicas == d.Status.ReadyReplicas && d.Status.ReadyReplicas == d.Status.AvailableReplicas {
```
[the remark this answers is not in the document]
Comment (kwoodson) [reply pkg/checks/checks.go]: added.

File: pkg/checks/checks.go (hunk 15 above)
Comment (kwoodson) [reply pkg/checks/checks.go]: Will remove replicaset testing.

**Comment by kwoodson** [comment]
@jim-minter, @kargakis,

With a latest rebase everything completed successfully.

```
+ az group deployment wait -g kwoodsonhcp -n azuredeploy --created --interval 10
+ KUBECONFIG=_data/_out/admin.kubeconfig
+ go run cmd/healthcheck/healthcheck.go
```

File: pkg/addons/client.go [hunk 26]
```diff
@@ -1,23 +1,17 @@
...
-	"net"
-	"net/http"
-	"net/url"
-	"os"
 	"reflect"
-	"syscall"
-	"time"
 
 	"github.com/go-test/deep"
+	"github.com/openshift/openshift-azure/pkg/checks"
```
Comment (jim-minter) [inline pkg/addons/client.go]: we seem to have the style of keeping the github.com/openshift/openshift-azure imports in a separate block - separate for consistency

File: pkg/checks/checks.go [hunk 27]
```diff
@@ -0,0 +1,169 @@
...
+	}
+
+	for {
+		resp, err := cli.Do(req)
+		if err, ok := err.(*url.Error); ok && (err.Timeout() || err.Err == io.EOF || err.Err == io.ErrUnexpectedEOF) {
+			time.Sleep(time.Second)
+			continue
+		}
+
+		if err == io.EOF {
```
Comment (jim-minter) [inline pkg/checks/checks.go]: the implementation changed upstream while your PR was in progress (see the `return wait.PollInfinite`...) and you'll need to merge my changes and yours.  The end result should include at least the `wait.PollInfinite`, the `syscall.ENETUNREACH` bit and your new `err == io.EOF` bit.

File: pkg/checks/checks.go [hunk 28]
```diff
@@ -0,0 +1,167 @@
...
+	cli := &http.Client{
+		Transport: transport,
+		Timeout:   10 * time.Second,
+	}
+
+	req, err := http.NewRequest("GET", urltocheck, nil)
+	if err != nil {
+		return err
+	}
+	return wait.PollInfinite(time.Second, func() (bool, error) {
```
Comment (jim-minter) [inline pkg/checks/checks.go]: sorry, my bad, should be `wait.PollUntil(time.Second, func() (bool, error) { ... }, ctx.Done())`

File: pkg/healthcheck/healthcheck.go [hunk 29]
```diff
@@ -10,11 +10,69 @@ import (
...
+	if err != nil {
+		return err
+	}
+
+	t, err := rest.TransportFor(restconfig)
+	if err != nil {
+		return err
+	}
+
+	ctxHealthz, cancel := context.WithTimeout(ctx, 5*time.Minute)
```
Comment (jim-minter) [inline pkg/healthcheck/healthcheck.go]: remove this, use the passed in ctx instead of ctxHealthz

File: pkg/healthcheck/healthcheck.go [hunk 30]
```diff
@@ -10,11 +10,69 @@ import (
...
+	if err != nil {
+		return err
+	}
+
+	appsclient, err := appsclient.NewForConfig(restconfig)
+	if err != nil {
+		return err
+	}
+	// Ensure that the pods in default are healthy
+	ctxPods, cancel := context.WithTimeout(ctx, 5*time.Minute)
```
Comment (jim-minter) [inline pkg/healthcheck/healthcheck.go]: remove this, use the passed in ctx instead of ctxPods

File: pkg/healthcheck/healthcheck.go (hunk 29 above)
Comment (kwoodson) [reply pkg/healthcheck/healthcheck.go]: done.

File: pkg/healthcheck/healthcheck.go (hunk 30 above)
Comment (kwoodson) [reply pkg/healthcheck/healthcheck.go]: done.
